File: Leetcode/LC-75/cpp-version/2-GCD-Strings.cpp
// https://leetcode.com/problems/greatest-common-divisor-of-strings/submissions/1205371531/?envType=study-plan-v2&envId=leetcode-75
// COMPLETE

#include <string>

class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        if (str1 + str2 != str2 + str1) {
            return "";
        }
        
        int len1 = str1.length();
        int len2 = str2.length();
        
        int gcd_length = gcd(len1, len2);
        
        return str1.substr(0, gcd_length);
    }
    
private:
    int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
};

File: Leetcode/LC-75/java-version/_2GCDStrings.java
// https://leetcode.com/problems/greatest-common-divisor-of-strings/?envType=study-plan-v2&envId=leetcode-75
// COMPLETE

import java.util.*;

public class _2GCDStrings {
    private static boolean checkDivides(String s, String t) {
        int repeatCount = s.length() / t.length();
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < repeatCount; i++) {
            sb.append(t);
        }
        return s.equals(sb.toString());
    }

    public static String gcdOfStrings(String str1, String str2) {
        int gcdLength = gcd(str1.length(), str2.length());
        
        String candidate = str1.substring(0, gcdLength);
        
        if (checkDivides(str1, candidate) && checkDivides(str2, candidate)) {
            return candidate;
        } else {
            return "";
        }
    }

    private static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }
}
